File: 프로그래머스/3/132266. 부대복귀/부대복귀.js
function bfs(graph, destination, n) {
    const distance = Array(n + 1).fill(-1);
    const queue = [destination];
    distance[destination] = 0;

    while (queue.length > 0) {
        const current = queue.shift();

        for (const next of graph[current]) {
            if (distance[next] === -1) {
                distance[next] = distance[current] + 1;
                queue.push(next);
            }
        }
    }

    return distance;
}

function solution(n, roads, sources, destination) {
    const graph = Array.from({ length: n + 1 }, () => []);
    roads.forEach(([v1, v2]) => {
        graph[v1].push(v2);
        graph[v2].push(v1);
    });

    const distance = bfs(graph, destination, n);
    return sources.map(source => distance[source]);
}

File: 프로그래머스/3/132266. 부대복귀/부대복귀.py
from collections import deque

def bfs(graph, destination, n):
    distance = [-1] * (n + 1)
    queue = deque([destination])
    distance[destination] = 0

    while queue:
        current = queue.popleft()

        for next_node in graph[current]:
            if distance[next_node] == -1:
                distance[next_node] = distance[current] + 1
                queue.append(next_node)

    return distance

def solution(n, roads, sources, destination):
    graph = [[] for _ in range(n + 1)]
    for v1, v2 in roads:
        graph[v1].append(v2)
        graph[v2].append(v1)

    distance = bfs(graph, destination, n)
    return [distance[start] for start in sources]